src/NHibernate/Type/GenericOrderedSetType.cs

```diff
@@ -24,9 +27,41 @@ public GenericOrderedSetType(string role, string propertyRef)
 		{
 		}
 
+		static ConstructorInfo SetConstructor
+		{
+			get
+			{
+				if (_setConstructor == null)
+				{
+					lock (typeof(GenericOrderedSetType<>))
```

Comment:
I don't think that anyone would make a lock on this type. Btw lock on a type is how `[MethodImpl(Synchronized)]` attribute is implemented. 

The reason for this field being lazy initialized because we cannot throw expetions from the static constructor - this will make the class unusable.